src/main/java/com/minecolonies/coremod/entity/ai/citizen/hundertaker/EntityAIWorkUndertaker.java

```diff
@@ -0,0 +1,517 @@
+            {
+                building.setLastGraveData((GraveData) ((TileEntityGrave) entity).getGraveData());
+                return EMPTY_GRAVE;
+            }
+            building.ClearCurrentGrave();
+        }
+
+        return WANDER;
+    }
+
+    /**
+     * The undertaker wander in the city, learning more about magic
+     *
+     * @return the next IAIState
+     */
+    @NotNull
+    private IAIState wander()
+    {
+        @Nullable final BuildingGraveyard building = getOwnBuilding();
+        if (building == null || building.getBuildingLevel() < 1)
+        {
+            worker.getCitizenData().setVisibleStatus(null);
+            return IDLE;
+        }
+
+        if(wanderPos == null)
+        {
+            final BlockPos newWanderPos = worker.getCitizenColonyHandler().getColony().getBuildingManager().
+                    getRandomBuilding(b -> b.getSchematicName() == "citizen" || b instanceof BuildingMysticalSite || b instanceof BuildingEnchanter);
+            if (newWanderPos != null)
+            {
+                wanderPos = newWanderPos;
+            }
+            return getState();
+        }
+
+        if (walkToBlock(wanderPos, 1))
+        {
+            return getState();
+        }
+
+        if(wanderPos == building.getPosition())
+        {
+            wanderPos = null;
+            worker.decreaseSaturationForContinuousAction();
+            worker.getCitizenData().getCitizenSkillHandler().addXpToSkill(getOwnBuilding().getSecondarySkill(), XP_PER_WANDER, worker.getCitizenData());
+        }
+        else
+        {
+            wanderPos = building.getPosition();
+        }
+
+        return IDLE;
+    }
+
+    private IAIState emptyGrave()
+    {
+        @Nullable final BuildingGraveyard buildingGraveyard = getOwnBuilding();
+
+        if (buildingGraveyard == null || checkForToolOrWeapon(ToolType.SHOVEL) || buildingGraveyard.getGraveToWorkOn() == null)
+        {
+            return IDLE;
+        }
+
+        worker.getCitizenData().setVisibleStatus(EMPTYING_ICON);
+        worker.getCitizenStatusHandler().setLatestStatus(new TranslationTextComponent("com.minecolonies.coremod.status.emptying"));
+        worker.setSprinting(true);
+        unequip();
+
+        @Nullable final BlockPos gravePos = buildingGraveyard.getGraveToWorkOn();
+        final TileEntity entity = world.getTileEntity(gravePos);
+        if (entity instanceof TileEntityGrave)
+        {
+            if (((TileEntityGrave) entity).isEmpty())
+            {
+                return TRY_RESURRECT;
+            }
+
+            if (worker.getInventoryCitizen().isFull())
+            {
+                return INVENTORY_FULL;
+            }
+
+            // Still moving to the block
+            if (walkToBlock(gravePos, 1))
+            {
+                return getState();
+            }
+
+            if(effortCounter < EFFORT_EMPTY_GRAVE)
+            {
+                worker.swingArm(Hand.MAIN_HAND);
```

Comment:
do we want to make sure to equip the shovel here?

Reply:
what is this for? delay between picking up items?

Reply:
The shovel should only be equipped during the state DIG_GRAVE and BURY_CITIZEN. The Undertaker still move his arms around during state EMPTY_GRAVE and TRY_RESURRECT to show something is happening to the player, but it didn't make sense to me to have the shovel in hand. For your second question, yes this is a delay in picking up items. While playtesting, it felt weird to me that the Undertaker came to the grave, pickup everything and left directly. I added this small delay to make the action of pickup up item seems like real work.